File: sdks/elixir/test/webhook_test.exs
defmodule HookSniff.WebhookTest do
  use ExUnit.Case, async: true

  alias HookSniff.Webhook

  @test_secret "whsec_" <> Base.encode64("test-secret-key-for-hmac")
  @test_body ~s({"event":"order.created","data":{"order_id":"12345"}})
  @test_msg_id "msg_test123"
  @test_timestamp System.system_time(:second)

  defp sign_payload(secret, msg_id, timestamp, body) do
    raw_secret =
      secret
      |> String.replace_prefix("whsec_", "")
      |> Base.decode64!()

    content = "#{msg_id}.#{timestamp}.#{body}"
    sig = :crypto.mac(:hmac, :sha256, raw_secret, content)
    "v1,#{Base.encode64(sig)}"
  end

  defp valid_headers(msg_id \\ nil, timestamp \\ nil) do
    mid = msg_id || @test_msg_id
    ts = timestamp || @test_timestamp
    %{
      "webhook-id" => mid,
      "webhook-timestamp" => to_string(ts),
      "webhook-signature" => sign_payload(@test_secret, mid, ts, @test_body)
    }
  end

  # Test 1: Valid signature verification
  test "valid signature returns parsed JSON" do
    headers = valid_headers()
    assert {:ok, payload} = Webhook.verify(@test_body, headers, @test_secret)
    assert payload["event"] == "order.created"
    assert payload["data"]["order_id"] == "12345"
  end

  # Test 2: Invalid signature
  test "invalid signature returns error" do
    headers = Map.put(valid_headers(), "webhook-signature", "v1,invalid_signature_here")
    assert {:error, %Webhook.VerificationError{}} = Webhook.verify(@test_body, headers, @test_secret)
  end

  # Test 3: Missing webhook-id header
  test "missing webhook-id returns error" do
    headers = Map.delete(valid_headers(), "webhook-id")
    assert {:error, %Webhook.VerificationError{}} = Webhook.verify(@test_body, headers, @test_secret)
  end

  # Test 4: Missing webhook-timestamp header
  test "missing webhook-timestamp returns error" do
    headers = Map.delete(valid_headers(), "webhook-timestamp")
    assert {:error, %Webhook.VerificationError{}} = Webhook.verify(@test_body, headers, @test_secret)
  end

  # Test 5: Missing webhook-signature header
  test "missing webhook-signature returns error" do
    headers = Map.delete(valid_headers(), "webhook-signature")
    assert {:error, %Webhook.VerificationError{}} = Webhook.verify(@test_body, headers, @test_secret)
  end

  # Test 6: Expired timestamp (replay protection)
  test "expired timestamp returns error" do
    old_ts = System.system_time(:second) - 600
    headers = valid_headers(@test_msg_id, old_ts)
    assert {:error, %Webhook.VerificationError{}} = Webhook.verify(@test_body, headers, @test_secret)
  end

  # Test 7: Svix-branded headers
  test "accepts svix-branded headers" do
    standard = valid_headers()
    svix_headers = %{
      "svix-id" => standard["webhook-id"],
      "svix-timestamp" => standard["webhook-timestamp"],
      "svix-signature" => standard["webhook-signature"]
    }
    assert {:ok, payload} = Webhook.verify(@test_body, svix_headers, @test_secret)
    assert payload["event"] == "order.created"
  end

  # Test 8: Multiple comma-separated signatures
  test "verifies with multiple comma-separated signatures" do
    sig = sign_payload(@test_secret, @test_msg_id, @test_timestamp, @test_body)
    multi_sig = "v1,wrong_sig," <> elem(String.split(sig, ",", parts: 2), 1)
    headers = Map.put(valid_headers(), "webhook-signature", multi_sig)
    assert {:ok, _} = Webhook.verify(@test_body, headers, @test_secret)
  end

  # Test 9: sign() produces verifiable signature
  test "sign() produces a verifiable signature" do
    sig = Webhook.sign(@test_msg_id, @test_timestamp, @test_body, @test_secret)
    assert String.starts_with?(sig, "v1,")

    headers = %{
      "webhook-id" => @test_msg_id,
      "webhook-timestamp" => to_string(@test_timestamp),
      "webhook-signature" => sig
    }
    assert {:ok, payload} = Webhook.verify(@test_body, headers, @test_secret)
    assert payload["event"] == "order.created"
  end

  # Test 10: Secret prefix handling
  test "works with and without whsec_ prefix" do
    raw_secret = Base.encode64("test-secret-key-for-hmac")
    headers = valid_headers()

    assert {:ok, _} = Webhook.verify(@test_body, headers, @test_secret)
    assert {:ok, _} = Webhook.verify(@test_body, headers, raw_secret)
  end

  # Test 11: Invalid timestamp format
  test "invalid timestamp returns error" do
    headers = Map.put(valid_headers(), "webhook-timestamp", "not_a_number")
    assert {:error, %Webhook.VerificationError{}} = Webhook.verify(@test_body, headers, @test_secret)
  end

  # Test 12: Returns raw string when payload is not JSON
  test "returns raw string for non-JSON payload" do
    raw_body = "not json"
    ts = System.system_time(:second)
    sig = sign_payload(@test_secret, @test_msg_id, ts, raw_body)

    headers = %{
      "webhook-id" => @test_msg_id,
      "webhook-timestamp" => to_string(ts),
      "webhook-signature" => sig
    }

    assert {:ok, "not json"} = Webhook.verify(raw_body, headers, @test_secret)
  end
end

File: sdks/php/test/WebhookTest.php
<?php

declare(strict_types=1);

namespace HookSniff\Test;

use HookSniff\Webhook;
use HookSniff\WebhookVerificationError;
use PHPUnit\Framework\TestCase;

class WebhookTest extends TestCase
{
    private string $testSecret;
    private string $testBody;
    private string $testMsgId;
    private int $testTimestamp;

    protected function setUp(): void
    {
        $this->testSecret = 'whsec_' . base64_encode('test-secret-key-for-hmac-verify');
        $this->testBody = '{"event":"order.created","data":{"order_id":"12345"}}';
        $this->testMsgId = 'msg_test123';
        $this->testTimestamp = time();
    }

    private function signPayload(string $secret, string $msgId, int $timestamp, string $body): string
    {
        $raw = str_starts_with($secret, 'whsec_') ? substr($secret, 6) : $secret;
        $decoded = base64_decode($raw, true) ?: $raw;
        $content = $msgId . '.' . $timestamp . '.' . $body;
        $hmac = base64_encode(hash_hmac('sha256', $content, $decoded, true));
        return 'v1,' . $hmac;
    }

    private function validHeaders(?string $msgId = null, ?int $timestamp = null): array
    {
        $mid = $msgId ?? $this->testMsgId;
        $ts = $timestamp ?? $this->testTimestamp;
        return [
            'webhook-id' => $mid,
            'webhook-timestamp' => (string) $ts,
            'webhook-signature' => $this->signPayload($this->testSecret, $mid, $ts, $this->testBody),
        ];
    }

    // Test 1: Valid signature verification
    public function testValidSignatureReturnsParsedJson(): void
    {
        $wh = new Webhook($this->testSecret);
        $result = $wh->verify($this->testBody, $this->validHeaders());
        $this->assertSame('order.created', $result['event']);
        $this->assertSame('12345', $result['data']['order_id']);
    }

    // Test 2: Invalid signature
    public function testInvalidSignatureThrowsError(): void
    {
        $wh = new Webhook($this->testSecret);
        $headers = $this->validHeaders();
        $headers['webhook-signature'] = 'v1,aW52YWxpZHNpZ25hdHVyZQ==';
        $this->expectException(WebhookVerificationError::class);
        $wh->verify($this->testBody, $headers);
    }

    // Test 3: Missing webhook-id header
    public function testMissingWebhookIdThrowsError(): void
    {
        $wh = new Webhook($this->testSecret);
        $headers = $this->validHeaders();
        unset($headers['webhook-id']);
        $this->expectException(WebhookVerificationError::class);
        $this->expectExceptionMessage('Missing webhook-id');
        $wh->verify($this->testBody, $headers);
    }

    // Test 4: Missing webhook-timestamp header
    public function testMissingWebhookTimestampThrowsError(): void
    {
        $wh = new Webhook($this->testSecret);
        $headers = $this->validHeaders();
        unset($headers['webhook-timestamp']);
        $this->expectException(WebhookVerificationError::class);
        $this->expectExceptionMessage('Missing webhook-timestamp');
        $wh->verify($this->testBody, $headers);
    }

    // Test 5: Missing webhook-signature header
    public function testMissingWebhookSignatureThrowsError(): void
    {
        $wh = new Webhook($this->testSecret);
        $headers = $this->validHeaders();
        unset($headers['webhook-signature']);
        $this->expectException(WebhookVerificationError::class);
        $this->expectExceptionMessage('Missing webhook-signature');
        $wh->verify($this->testBody, $headers);
    }

    // Test 6: Expired timestamp (replay protection)
    public function testExpiredTimestampThrowsError(): void
    {
        $wh = new Webhook($this->testSecret);
        $oldTimestamp = time() - 600; // 10 minutes ago
        $headers = $this->validHeaders($this->testMsgId, $oldTimestamp);
        $this->expectException(WebhookVerificationError::class);
        $this->expectExceptionMessage('too old or too new');
        $wh->verify($this->testBody, $headers);
    }

    // Test 7: Svix-branded headers work
    public function testSvixBrandedHeadersWork(): void
    {
        $wh = new Webhook($this->testSecret);
        $ts = $this->testTimestamp;
        $sig = $this->signPayload($this->testSecret, $this->testMsgId, $ts, $this->testBody);
        $headers = [
            'svix-id' => $this->testMsgId,
            'svix-timestamp' => (string) $ts,
            'svix-signature' => $sig,
        ];
        $result = $wh->verify($this->testBody, $headers);
        $this->assertSame('order.created', $result['event']);
    }

    // Test 8: Multiple comma-separated signatures
    public function testMultipleSignaturesWork(): void
    {
        $wh = new Webhook($this->testSecret);
        $ts = $this->testTimestamp;
        $realSig = $this->signPayload($this->testSecret, $this->testMsgId, $ts, $this->testBody);
        $combinedSig = 'v1,aW52YWxpZA==,' . $realSig;
        $headers = $this->validHeaders();
        $headers['webhook-signature'] = $combinedSig;
        $result = $wh->verify($this->testBody, $headers);
        $this->assertSame('order.created', $result['event']);
    }

    // Test 9: sign() produces verifiable signature
    public function testSignProducesVerifiableSignature(): void
    {
        $wh = new Webhook($this->testSecret);
        $ts = $this->testTimestamp;
        $sig = $wh->sign($this->testMsgId, $ts, $this->testBody);
        $this->assertStringStartsWith('v1,', $sig);

        $headers = [
            'webhook-id' => $this->testMsgId,
            'webhook-timestamp' => (string) $ts,
            'webhook-signature' => $sig,
        ];
        $result = $wh->verify($this->testBody, $headers);
        $this->assertSame('order.created', $result['event']);
    }

    // Test 10: Secret with and without whsec_ prefix
    public function testSecretWithAndWithoutPrefix(): void
    {
        $rawSecret = base64_encode('test-secret-key-for-hmac-verify');
        $prefixed = 'whsec_' . $rawSecret;

        $whPrefixed = new Webhook($prefixed);
        $whRaw = new Webhook($rawSecret);

        $ts = $this->testTimestamp;
        $sig = $whRaw->sign($this->testMsgId, $ts, $this->testBody);
        $headers = [
            'webhook-id' => $this->testMsgId,
            'webhook-timestamp' => (string) $ts,
            'webhook-signature' => $sig,
        ];

        $result1 = $whPrefixed->verify($this->testBody, $headers);
        $result2 = $whRaw->verify($this->testBody, $headers);
        $this->assertSame('order.created', $result1['event']);
        $this->assertSame('order.created', $result2['event']);
    }

    // Test 11: Invalid timestamp format
    public function testInvalidTimestampFormatThrowsError(): void
    {
        $wh = new Webhook($this->testSecret);
        $headers = $this->validHeaders();
        $headers['webhook-timestamp'] = 'not_a_number';
        $this->expectException(WebhookVerificationError::class);
        $this->expectExceptionMessage('Invalid webhook-timestamp');
        $wh->verify($this->testBody, $headers);
    }

    // Test 12: verifyWithSecret static convenience
    public function testVerifyWithSecretStaticMethod(): void
    {
        $ts = $this->testTimestamp;
        $sig = $this->signPayload($this->testSecret, $this->testMsgId, $ts, $this->testBody);
        $headers = [
            'webhook-id' => $this->testMsgId,
            'webhook-timestamp' => (string) $ts,
            'webhook-signature' => $sig,
        ];
        $result = Webhook::verifyWithSecret($this->testSecret, $this->testBody, $headers);
        $this->assertSame('order.created', $result['event']);
    }

    // Test 13: Non-JSON payload returns raw string
    public function testNonJsonPayloadReturnsRawString(): void
    {
        $wh = new Webhook($this->testSecret);
        $rawBody = 'not json';
        $ts = $this->testTimestamp;
        $sig = $this->signPayload($this->testSecret, $this->testMsgId, $ts, $rawBody);
        $headers = [
            'webhook-id' => $this->testMsgId,
            'webhook-timestamp' => (string) $ts,
            'webhook-signature' => $sig,
        ];
        $result = $wh->verify($rawBody, $headers);
        $this->assertSame('not json', $result);
    }
}
